### PR TITLE
Initial Commit

### PR DESCRIPTION
**Pull request by ZER0**

### CODE REVIEW

File: src/decode.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
+use crate::{Error, JubJubAffine, JubJubScalar};
+use subtle::CtOption;
+
+// TODO: move this logic (in this form or another) to jubjub repo, or higher
+// repo than `dusk-pki`
+
+pub trait From32Bytes {
+    fn from_bytes(bytes: [u8; 32]) -> CtOption<Self>
```
Comment (vlopes11) [inline src/decode.rs:8]: `CtOption` is very unstable and lack most of the benefits from the standard `Option`. If we are to use that as standard only because of conditional selection, then maybe we should think on a solution to encapsulate the conditional selection instead of setting subtle as our standard.

File: src/spend/public.rs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    /// Gets `B` (`b·G`)
+    pub fn B(&self) -> &JubJubExtended {
+        &self.B
+    }
+
+    /// Generates new `PKr = H(A · r) · G + B` from a given `r`
+    pub fn gen_stealth_address(&self, r: &JubJubScalar) -> StealthAddress {
+        let G = GENERATOR.to_niels();
```
Comment (vlopes11) [inline src/spend/public.rs]: Any particular reason to use niels here?

File: src/spend/secret.rs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    pub fn from_rng<R: RngCore + CryptoRng>(rng: &mut R) -> Self {
+        let a = JubJubScalar::random(rng);
+        let b = JubJubScalar::random(rng);
+
+        SecretKey::new(a, b)
+    }
+
+    /// Derive the secret to deterministically construct a [`PublicKey`]
+    pub fn public_key(&self) -> PublicKey {
+        let A = JubJubExtended::from(GENERATOR).mul(&self.a);
```
Comment (vlopes11) [inline src/spend/secret.rs]: Use `jubjub::GENERATOR_EXTENDED` instead (as in the occurrences below)

File: src/spend/stealth.rs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    fn stealth_address<'a>(&'a self) -> &'a StealthAddress;
+}
+
+//. To obfuscate the identity of the participants, we utilizes a Stealth Address
+//. system.
+/// A `StealthAddress` is composed by a one-time public key (`pk_r`, the actual
+// address) and a random point `R`.
+#[derive(Debug, Clone, Copy)]
+pub struct StealthAddress {
+    pub(crate) R: JubJubExtended,
```
Comment (vlopes11) [inline src/spend/stealth.rs:20]: Why public for the crate if we have the getters?

File: src/sponge.rs [hunk 5]
```diff
@@ -0,0 +1,7 @@
+use crate::{JubJubAffine, JubJubExtended, JubJubScalar};
+use poseidon252::sponge::sponge::sponge_hash;
+
+pub fn hash(p: &JubJubExtended) -> JubJubScalar {
+    let p = JubJubAffine::from(p);
+    JubJubScalar::from_raw(sponge_hash(&[p.get_x(), p.get_y()]).reduce().0)
```
Comment (vlopes11) [inline src/sponge.rs]: Why sponge mode if its easier to integrate into zk circuit a normal permutation and the width of 5 accepts these two inputs?

**Review by vlopes11 — CHANGES_REQUESTED** [review]

File: src/spend/public.rs (hunk 2 above)
Comment (ZER0) [reply src/spend/public.rs:40]: Because it's the idiomatic way to use the `GENERATOR` in the multiplication (in accordance with the jubjub crate) and it's consistent with the rest of our code base (@CPerezz used that as well in other repo).

File: src/decode.rs (hunk 1 above)
Comment (ZER0) [reply src/decode.rs:8]: I'm not using `CtOption` because I want to, it's because the `from_bytes` impl of JubJub's return it, so the trait needs to be consistent with that.

AFAIK @CPerezz is working on a PR for the subtle crate that will implement a conversion from `CtOption` to `Option`. So basically this is just a temporary measure.

File: src/spend/public.rs (hunk 2 above)
Comment (ZER0) [reply src/spend/public.rs]: Just to perform the `GENERATOR` multiplication – since we have the same code in other part of our code base.
If that is not the way to go we probably should refactor everything in order to be consistent.

File: src/spend/secret.rs (hunk 3 above)
Comment (ZER0) [reply src/spend/secret.rs]: This is actually a leftover. It should be `GENERATOR.to_niels()`.
Is the `jubjub::GENERATOR_EXTENDED` a recent impl? I didn't had it in jubjub when I started the code. I will move the code to that then.

File: src/spend/stealth.rs (hunk 4 above)
Comment (ZER0) [reply src/spend/stealth.rs:20]: Just to ensure the optimization of the crate, but it's indeed a bit of overkill. I can remove it.

File: src/sponge.rs (hunk 5 above)
Comment (ZER0) [reply src/sponge.rs]: Not sure I'm following, sorry. Could you rephrase?
Take in account that this doesn't need to be proved in zk, it's just an optimization to the previous hashing (and I double checked with Dmitry that the approach was sound).

Basically we want to hash an `AffinePoint` into a jubjub's scalar.  Since Poseidon takes BLS scalars, and the AffinePoint coords are BLS, it's easy doing so. The problem was the returned value that could be not a valid jubjub's scalar – but the function take care of it thanks to the `from_raw` + `reduce`.

File: src/spend/stealth.rs (hunk 4 above)
Comment (ZER0) [reply src/spend/stealth.rs:20]: Actually no, I forgot why I added in the first place: because currently (by design) it shouldn't be possible instantiate directly `StealthAddress` from outside the `dusk-pki` crates. That's why the visibility `pub(crate)`. Since also there are no other "constructor" (such `new`), the only way to have a `StealthAddress` is to use `gen_stealth_address` from a `PublicSpendKey`.

File: src/decode.rs (hunk 1 above)
Comment (CPerezz) [reply src/decode.rs:8]: Yes, I'll try to do it today at night or so. Henry already agreed on the idea I had. So it's a matter of merge the PR and then wait for a release to have this conversion finally avaliable.

File: src/errors.rs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    InvalidParameters,
+    /// Bad Length
+    #[error("Bad Length (expected {expected:?}, got {found:?})")]
+    BadLength {
+        /// The found length
+        found: usize,
+        /// The expected length
+        expected: usize,
+    },
+    // TODO: this should be handled better, it's too generic
```
Comment (CPerezz) [inline src/errors.rs:21]: As I said we can convert the errors with `into()`. Maybe open an isue in `anyhow` and ask for this specific situation possible solutions would be nice.

File: src/spend/secret.rs [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+pub struct SecretKey {
+    a: JubJubScalar,
+    b: JubJubScalar,
+}
+
+impl Default for SecretKey {
+    fn default() -> Self {
+        SecretKey {
+            a: JubJubScalar::random(&mut rand::thread_rng()),
+            b: JubJubScalar::random(&mut rand::thread_rng()),
```
Comment (CPerezz) [inline src/spend/secret.rs]: I've never seen `default` used in this way. Usually in `default` you return the default impl of your attributes. Since when you call default you usually want to generate the structure to edit it later. So it doesn't make any sense to waste time computing useless scalars.

If you want to provide a function that does this, my suggestion would be:

```rust
pub fn random<T>(rand: &mut T) -> SecretKey
    where
        T: Rng + CryptoRng,
    {
        SecretKey {
            a: JubJubScalar::random(rand),
            b: JubJubScalar::random(rand),
        }
    }
```

Comment (CPerezz) [reply src/spend/secret.rs]: I just saw that you already have it. Therefore, this should be something like:
```rust
    fn default() -> Self {
        SecretKey {
            a: JubJubScalar::default(),
            b: JubJubScalar::default(),
        }
    }
```

File: src/spend/secret.rs [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    /// Generate a `sk_r = H(a · R) + b`
+    pub fn sk_r(&self, sa: &StealthAddress) -> JubJubScalar {
+        let aR = sa.R() * self.a;
+        let aR = sponge::hash(&aR);
+        aR + self.b
+    }
+
+    /// Deterministically create a new [`SecretKey`] from a random number
+    /// generator
+    pub fn from_rng<R: RngCore + CryptoRng>(rng: &mut R) -> Self {
```
Comment (CPerezz) [inline src/spend/secret.rs]: The convention is to call it `random` but it's just an idea. So this is also fine if you want to keep it.

File: src/spend/secret.rs [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        let mut hasher = Sha256::default();
+        hasher.input(bytes);
+        let bytes = hasher.result();
+
+        let mut seed = [0u8; 32];
+        seed.copy_from_slice(&bytes[..32]);
+
+        SecretKey::from_rng(&mut StdRng::from_seed(seed))
+    }
+}
```
Comment (CPerezz) [inline src/spend/secret.rs]: Is this necessary?

File: src/sponge.rs (hunk 5 above)
Comment (CPerezz) [reply src/sponge.rs]: The approach is fine for now. But definitely we need to give it a thought on the future.

File: src/sponge.rs [hunk 10]
```diff
@@ -0,0 +1,7 @@
+use crate::{JubJubAffine, JubJubExtended, JubJubScalar};
+use poseidon252::sponge::sponge::sponge_hash;
+
+pub fn hash(p: &JubJubExtended) -> JubJubScalar {
```
Comment (CPerezz) [inline src/sponge.rs:5]: I think it's important to document here that we need to hash the Affine representation, and why is it done like this. Since on this way on the future no one else will have the same issue you had @ZER0

**Review by CPerezz — CHANGES_REQUESTED** [review]

File: src/errors.rs (hunk 6 above)
Comment (ZER0) [reply src/errors.rs:21]: `into()` is not necessary, it's implicit with `?`. The issue is that you need to explicitly add the conversion `From` trait when you want to convert from (for example) anyhow / thiserror to `io::Error`. 
For the other way around, you can use `#[from]` for `thiserror`.

Notice that this problem arise only when the function signature requires `io::Error`, otherwise you could just have an `anyhow::Error` as return value.

File: src/spend/secret.rs (hunk 7 above)
Comment (ZER0) [reply src/spend/secret.rs]: This was the original implementation of `Phoenix` for the `SecretKey`. I kept because it actually made sense to me.
Notice that the fact that "Usually in `default` you return the default impl of your attributes" is a false assumption. By Rust specs, "*default values are often some kind of initial value, identity value, or anything else that may make sense as a default.*"

In case of the `SecretSpendKey` the only default values that makes sense are random scalars. If you don't want to construct a "default" `SecretSpendKey`, then you pass such values using the constructor `new`, or you construct it from random bytes.

File: src/spend/secret.rs (hunk 8 above)
Comment (ZER0) [reply src/spend/secret.rs]: I don't mind, I just left the original name, I can replace with `random`, sounds good to me.

File: src/spend/secret.rs (hunk 9 above)
Comment (ZER0) [reply src/spend/secret.rs]: Care to elaborate?

File: src/sponge.rs (hunk 10 above)
Comment (ZER0) [reply src/sponge.rs:5]: You mean the line below, not this one, where I had to convert the Extended Point to Affine Point? I'll do that.

File: src/spend/secret.rs (hunk 7 above)
Comment (ZER0) [reply src/spend/secret.rs]: Of course as soon as a scenario that invalidate this statements arise, we can definitely change it.
But I don't see atm something like that for the `SecretSpendKey` (It's not like the `PublicSpendKey`).

File: src/errors.rs (hunk 6 above)
Comment (CPerezz) [reply src/errors.rs:21]: Ye, I see. It's fine then. Not something we should loose time with now. Resolved.

File: src/spend/secret.rs (hunk 7 above)
Comment (CPerezz) [reply src/spend/secret.rs]: It's fine then.

File: src/spend/secret.rs (hunk 9 above)
Comment (CPerezz) [reply src/spend/secret.rs]: It might be confusing since it can seem that it's intended to read bytes and from there get the key (as if it was a `Read` trait impl). If you want to get a random `SecretKey` just use the `from_rng` method.

File: src/sponge.rs (hunk 10 above)
Comment (CPerezz) [reply src/sponge.rs:5]: Exactly. I think either inside or as a fn comment. This needs to be commented.

File: src/spend/secret.rs (hunk 8 above)
Comment (CPerezz) [reply src/spend/secret.rs]: Whatever you prefer. All of the other repos we use have `random` so probably makes sense to stick with it and be consistent.

File: src/spend/secret.rs (hunk 9 above)
Comment (ZER0) [reply src/spend/secret.rs]: IMVHO it's not confusing because in order to read bytes to generate a `SecretKey` you don't do that from a slice of bytes. As we do for all our other types, you do it from a fixed size array.

In this case, it's not about having a random `SecretKey`, is about generating a `SecretKey` by arbitrary bytes. So, as you can see, the scope is different.

If your question is _"Do we have the need to generate a `SecretKey` from arbitrary bytes?"_ I will delegate the question to Togh. The original implementation has this feature, so I bring it with me in Dusk PKI.

Comment (CPerezz) [reply src/spend/secret.rs]: If the original impl has this feature I said nothing. Even if it looks or not confusing to me, the specs are the specs. So this LGTM therefore.

**Review by CPerezz — APPROVED** [review]
LGTM
